ns/Cns_setsegattrs.man
.\" Copyright (C) 1999-2002 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH CNS_SETSEGATTRS 3 "$Date: 2009/07/23 12:22:04 $" CASTOR "Cns Library Functions"
.SH NAME
Cns_setsegattrs \- set file segment attributes
.SH SYNOPSIS
.B #include <sys/types.h>
.br
\fB#include "Cns_api.h"\fR
.sp
.BI "int Cns_setsegattrs (const char *" path ,
.BI "struct Cns_fileid *" file_uniqueid ,
.BI "int " nbseg ,
.BI "struct Cns_segattrs *" segattrs,
.BI "time_t " last_mod_time )
.SH DESCRIPTION
.B Cns_setsegattrs
sets the file segment(s) attributes. Takes care that the file was not modified by somebody else in the mean time.
If it was, ignores the request and fails with ENSFILECHG.
This function should be called by the stager to set segment size, tape
visual identifier and file sequence number or blockid when all segments
which belong to a file have been written to tape.
The file can be identified by
.I path
name or by
.IR file_uniqueid .
If both are specified,
.I file_uniqueid
is used.
.TP
.I path
specifies the logical pathname relative to the current CASTOR directory or
the full CASTOR pathname.
.TP
.I nbseg
the number of segments in the segattrs array.
.TP
.I segattrs
is a pointer to an array of Cns_segattrs structures provided by the application.
.PP
.nf
.ft CW
struct Cns_segattrs {
        int             copyno;
        int             fsec;           /* file section number */
        u_signed64      segsize;        /* file section size */
        int             compression;    /* compression factor */
        char            s_status;       /* 'D' --> deleted */
        char            vid[CA_MAXVIDLEN+1];
        int             side;
        int             fseq;           /* file sequence number */
        unsigned char   blockid[4];     /* for positionning with locate */
        char            checksum_name[CA_MAXCKSUMNAMELEN+1];
        unsigned long   checksum;
};
.ft
.fi
.TP
.I copyno
is the copy number. If specified as zero, nsd will compute the actual
value.
.TP
.I fsec
is the file segment number, It is 1 for single volume files and >= 1 for files
spanning over several tapes.
.TP
.I last_mod_time
specifies the last modification time of the file, as known by the caller. This is used to double check that the file was not modified concurrently by somebody else.
.SH RETURN VALUE
This routine returns 0 if the operation was successful or -1 if the operation
failed. In the latter case,
.B serrno
is set appropriately.
.SH ERRORS
.TP 1.3i
.B ENOENT
The referenced file does not exist anymore.
.TP
.B EACCES
Search permission is denied on a component of the
.I path
prefix or the caller effective user ID does not match the owner ID of the file
or write permission on the file itself is denied.
.TP
.B ENSFILECHG
The file has been changed/overwritten by somebody else since the last_mod_time given. Request was thus ignored.
.TP
.B ENSCLASSNOSEGS
File class does not allow a copy on tape. The maximum number of copies of the
file on tape, defined by the file class, prevents the creation of segments.
.TP
.B EFAULT
.I path
and
.I file_uniqueid
or
.I vid
are NULL pointers.
.TP
.B ENOTDIR
A component of
.I path
prefix is not a directory.
.TP
.B EISDIR
The file is not a regular file.
.TP
.B EINVAL
The length of
.I vid
exceeds
.BR CA_MAXVIDLEN .
.TP
.B ENAMETOOLONG
The length of
.I path
exceeds
.B CA_MAXPATHLEN
or the length of a
.I path
component exceeds
.BR CA_MAXNAMELEN .
.TP
.B SENOSHOST
Host unknown.
.TP
.B SENOSSERV
Service unknown.
.TP
.B SECOMERR
Communication error.
.TP
.B ENSNACT
Name server is not running or is being shutdown.
.SH SEE ALSO
.BR Castor_limits(4) ,
.BR Cns_chdir(3) ,
.BR Cns_getsegattrs(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


ns/Cns_setsegattrs.c
/*
 * Copyright (C) 1999-2002 by CERN/IT/PDP/DM
 * All rights reserved
 */

/* Cns_setsegattrs - set file segments attributes */

#include <errno.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>
#include "marshall.h"
#include "Cns_api.h"
#include "Cns.h"
#include "serrno.h"
#include <string.h>

int
Cns_setsegattrs(const char *path, struct Cns_fileid *file_uniqueid, int nbseg, struct Cns_segattrs *segattrs, time_t last_mod_time)
{
  char *actual_path;
  int c;
  char func[16];
  gid_t gid;
  int i;
  int msglen;
  char *q;
  char *sbp;
  char sendbuf[REQBUFSZ];
  char server[CA_MAXHOSTNAMELEN+1];
  struct Cns_api_thread_info *thip;
  uid_t uid;
  u_signed64 zero = 0;

  strncpy (func, "Cns_setsegattrs", 16);
  if (Cns_apiinit (&thip))
    return (-1);
  Cns_getid(&uid, &gid);

  if ((! path && ! file_uniqueid) || ! segattrs) {
    serrno = EFAULT;
    return (-1);
  }

  if (path && strlen (path) > CA_MAXPATHLEN) {
    serrno = ENAMETOOLONG;
    return (-1);
  }

  if (file_uniqueid && *file_uniqueid->server)
    strcpy (server, file_uniqueid->server);
  else
    if (Cns_selectsrvr (path, thip->server, server, &actual_path))
      return (-1);

  /* Build request header */

  sbp = sendbuf;
  marshall_LONG (sbp, CNS_MAGIC5);
  marshall_LONG (sbp, CNS_SETSEGAT);
  q = sbp;        /* save pointer. The next field will be updated */
  msglen = 3 * LONGSIZE;
  marshall_LONG (sbp, msglen);

  /* Build request body */

  marshall_LONG (sbp, uid);
  marshall_LONG (sbp, gid);
  marshall_HYPER (sbp, thip->cwd);
  if (file_uniqueid && *file_uniqueid->server) {
    marshall_HYPER (sbp, file_uniqueid->fileid);
    marshall_STRING (sbp, "");
  } else {
    marshall_HYPER (sbp, zero);
    marshall_STRING (sbp, actual_path);
  }
  marshall_TIME_T (sbp, last_mod_time);
  marshall_WORD (sbp, nbseg);
  for (i = 0; i < nbseg; i++) {
    marshall_WORD (sbp, (segattrs+i)->copyno);
    marshall_WORD (sbp, (segattrs+i)->fsec);
    marshall_HYPER (sbp, (segattrs+i)->segsize);
    marshall_LONG (sbp, (segattrs+i)->compression);
    marshall_BYTE (sbp, (segattrs+i)->s_status);
    marshall_STRING (sbp, (segattrs+i)->vid);
    marshall_WORD (sbp, (segattrs+i)->side);
    marshall_LONG (sbp, (segattrs+i)->fseq);
    marshall_OPAQUE (sbp, (segattrs+i)->blockid, 4);
    marshall_STRING (sbp, (segattrs+i)->checksum_name);
    marshall_LONG (sbp, (segattrs+i)->checksum);
  }

  msglen = sbp - sendbuf;
  marshall_LONG (q, msglen); /* update length field */

  c = send2nsd (NULL, server, sendbuf, msglen, NULL, 0);
  if (c && serrno == SENAMETOOLONG) serrno = ENAMETOOLONG;
  return (c);
}
